zhaquirks/sinope/switch.py

```diff
@@ -56,36 +103,114 @@ class TankSize(t.enum8):
+        M_90 = 0x1518
+        H_3 = 0x2A30
+        H_6 = 0x5460
+        H_12 = 0xA8C0
+        H_24 = 0x15180
 
     cluster_id = SINOPE_MANUFACTURER_CLUSTER_ID
     name = "Sinopé Manufacturer specific"
     ep_attribute = "sinope_manufacturer_specific"
     attributes = {
         0x0002: ("keypad_lockout", KeypadLock, True),
+        0x0003: ("firmware_number", t.uint16_t, True),
         0x0004: ("firmware_version", t.CharacterString, True),
+        0x0010: ("outdoor_temp", t.int16s, True),
         0x0013: ("tank_size", TankSize, True),
         0x0060: ("connected_load", t.uint16_t, True),
         0x0070: ("current_load", t.bitmap8, True),
         0x0076: ("dr_config_water_temp_min", t.uint8_t, True),
         0x0077: ("dr_config_water_temp_time", t.uint8_t, True),
         0x0078: ("dr_wt_time_on", t.uint16_t, True),
+        0x0090: ("current_summation_delivered", t.uint32_t, True),
         0x00A0: ("timer", t.uint32_t, True),
+        0x00A1: ("timer_countdown", t.uint32_t, True),
+        0x0200: ("status", t.bitmap32, True),
+        0x0230: ("alarm_flow_threshold", FlowAlarm, True),
+        0x0231: ("alarm_options", AlarmAction, True),
+        0x0240: ("flow_meter_config", Array, True),
+        0x0241: ("valve_countdown", t.uint32_t, True),
+        0x0250: ("power_source", PowerSource, True),
+        0x0251: ("emergency_power_source", EmergencyPower, True),
+        0x0252: ("abnormal_flow_duration", FlowDuration, True),
+        0x0253: ("abnormal_flow_action", AbnormalAction, True),
         0x0283: ("cold_load_pickup_status", ColdStatus, True),
         0xFFFD: ("cluster_revision", t.uint16_t, True),
     }
 
 
+class CustomBasicCluster(CustomCluster, Basic):
+    """Custom Basic Cluster."""
+
+    class PowerSource(t.enum8):
+        """Power source."""
+
+        Unknown = 0x0000
+        Battery = 0x0003
+        DC_source = 0x0004
+        ACUPS_01 = 0x0081
+        ACUPS01 = 0x0082
+
+    attributes = Basic.attributes.copy()
+    attributes.update(
+        {
+            0x0007: ("power_source", PowerSource, True),
+        }
+    )
+
+
 class CustomMeteringCluster(CustomCluster, Metering):
     """Custom Metering Cluster."""
 
+    class ValveStatus(t.bitmap8):
+        """valve_status."""
+
+        Off = 0x00
+        Off_armed = 0x01
+        On = 0x02
+
+    class UnitOfMeasure(t.enum8):
+        """unit_of_measure."""
+
+        KWh = 0x00
+        Lh = 0x07
+
     DIVISOR = 0x0302
     _CONSTANT_ATTRIBUTES = {DIVISOR: 1000}
 
+    attributes = Metering.attributes.copy()
+    attributes.update(
+        {
+            0x0200: ("status", ValveStatus, True),
+            0x0300: ("unit_of_measure", UnitOfMeasure, True),
+        }
+    )
+
+
+class CustomDeviceTemperatureCluster(CustomCluster, DeviceTemperature):
+    """Custom DeviceTemperature Cluster."""
+
+    def _update_attribute(self, attrid, value):
+        if attrid == 0x0000:
+            super()._update_attribute(attrid, value * 100)
```

Comment:
Ideally, we'd also have a very small test for this.

It could just call `update_attribute` (forwards to `_update_attribute` in zigpy) on this cluster and then check if it's multiplied correctly.

As somewhat of an example, see this:

important part:
https://github.com/zigpy/zha-device-handlers/blob/fb794323fb11014c41843bebc5825bc9004cd23a/tests/test_xiaomi.py#L582-L589

first part:
https://github.com/zigpy/zha-device-handlers/blob/fb794323fb11014c41843bebc5825bc9004cd23a/tests/test_xiaomi.py#L528-L539

Otherwise, I can also see if I get around to creating this short test in (hopefully) the next few days.

Reply:
Resolved with added test in https://github.com/zigpy/zha-device-handlers/pull/2759/commits/3ec99ba6ab734db14c8c19a1d052a8e9a5cced42